src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,159 @@
+package duke;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.List;
+
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.ToDo;
+
+
+
+/**
+ * Storage links the TaskList from Duke to the user's local storage.
+ */
+public class Storage {
+    private final Path pathToStorage;
+    private List<String> allTasks;
+
+    /**
+     * Constructor to create a Storage object, which will be used when Duke creates, updates, or deletes Tasks.
+     *
+     * @param pathToStorage the path to the user's storage where the Tasks will be stored.
+     */
+    public Storage(Path pathToStorage) {
+        this.pathToStorage = pathToStorage;
+        try {
+            // Create directory if needed
+            Path parentPath = pathToStorage.getParent();
+            Files.createDirectories(parentPath);
+
+            if (!Files.exists(pathToStorage)) {
+                Files.createFile(pathToStorage);
+            }
+            // Get stored info if any
+            this.allTasks = Files.readAllLines(pathToStorage);
+        } catch (IOException e) {
+            System.out.println("Can't read file " + e.getMessage());
+        }
+    }
+
+    /**
+     * Converts string from task storage file to a task list for Duke to use.
+     *
+     * @return a list of Tasks readable by duke.Duke.
+     */
+    public ArrayList<Task> getAllTasks() throws DukeException {
+        ArrayList<Task> taskList = new ArrayList<>();
+        // Convert string to Tasks
+        for (String taskString : this.allTasks) {
+            char taskType = taskString.charAt(0);
+            String taskDetails = taskString.substring(taskString.lastIndexOf("|") + 2);
+            boolean isDone = false;
+
+            if (taskString.charAt(4) == '1') {
+                isDone = true;
+            }
+
+            switch (taskType) {
+            case 'T':
+                taskList.add(new ToDo(taskDetails, isDone));
+                break;
+            case 'E':
+                String eventDescription = taskString.substring(8, taskString.lastIndexOf("|") - 1);
+                taskList.add(new Event(eventDescription, isDone, formatTaskDateTime(taskDetails)));
+                break;
+            case 'D':
+                String deadlineDescription = taskString.substring(8, taskString.lastIndexOf("|") - 1);
+                taskList.add(new Deadline(deadlineDescription, isDone, formatTaskDateTime(taskDetails)));
+                break;
+            default:
+                System.out.println("Unable to determine type of task");
+                break;
+            }
+        }
+
+        return taskList;
+    }
+
+    private String formatTaskDateTime(String dateTime) {
```

Comment:
not a violation but - i saw that you used format later but just wanted to remind that DateTimeFormatter can format the output string so there is no need to format manually!